beater/request/context.go

```diff
@@ -0,0 +1,165 @@
+// under the License.
+
+package request
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/apm-server/beater/headers"
+	logs "github.com/elastic/apm-server/log"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	mimeTypeAny             = "*/*"
+	mimeTypeApplicationJSON = "application/json"
+)
+
+var (
+	mimeTypesJSON = []string{mimeTypeAny, mimeTypeApplicationJSON}
+)
+
+// Context abstracts request and response information for http requests
+type Context struct {
+	Req *http.Request
+
+	statusCode int
+	err        interface{}
+	stacktrace string
+
+	w http.ResponseWriter
+}
+
+// Reset allows to reuse a context by removing all request specific information
+func (c *Context) Reset(w http.ResponseWriter, r *http.Request) {
+	c.Req = r
+
+	c.w = w
+	c.statusCode = http.StatusOK
+	c.err = ""
+	c.stacktrace = ""
+}
+
+// Header returns the http.Header of the context's writer
+func (c *Context) Header() http.Header {
+	return c.w.Header()
+}
+
+// StatusCode returns the context's status code
+func (c *Context) StatusCode() int {
+	return c.statusCode
+}
+
+// Error returns the context's error
+func (c *Context) Error() interface{} {
+	return c.err
+}
+
+// Stacktrace returns the context's stacktrace
+func (c *Context) Stacktrace() string {
+	return c.stacktrace
+}
+
+// AddStacktrace sets a stacktrace for the context
+func (c *Context) AddStacktrace(stacktrace string) {
+	c.stacktrace = stacktrace
+}
+
+//TODO: All of the below methods will be changed during the response handling refactoring
+
+// WriteHeader sets status code in context and call the http.ResponseWriter method
+func (c *Context) WriteHeader(statusCode int) {
+	c.w.WriteHeader(statusCode)
+	c.statusCode = statusCode
+}
+
+// SendNotFoundErr is moved from the rootHandler and will be removed in the following refactoring
+func (c *Context) SendNotFoundErr() {
+	c.statusCode = http.StatusNotFound
+	http.NotFound(c.w, c.Req)
+}
+
+// Send is taking care of writing the response
+func (c *Context) Send(body interface{}, statusCode int) {
+	c.SendError(body, nil, statusCode)
+}
+
+// SendError sets and error and writes the response
+func (c *Context) SendError(body, err interface{}, statusCode int) {
+	c.err = err
+	c.w.Header().Set(headers.XContentTypeOptions, "nosniff")
+
+	if body == nil {
+		c.WriteHeader(statusCode)
+		return
+	}
+
+	if c.acceptJSON() {
+		c.sendJSON(body, statusCode)
+	} else {
+		c.sendPlain(body, statusCode)
+	}
+}
+
+func (c *Context) sendJSON(body interface{}, statusCode int) {
+	c.Header().Set(headers.ContentType, "application/json")
+	c.WriteHeader(statusCode)
+	buf, err := json.MarshalIndent(body, "", "  ")
+	if err != nil {
+		logp.NewLogger(logs.Response).Errorf("Error while generating a JSON error response: %v", err)
+		c.sendPlain(body, statusCode)
+		return
+	}
+
+	buf = append(buf, "\n"...)
+	n, _ := c.w.Write(buf)
+	c.w.Header().Set(headers.ContentLength, strconv.Itoa(n))
```

Comment:
isn't the body already written at this point?

Reply:
Yes, that is just moved code, as this PR doesn't intend to change or improve anything related to writing response. It only introduced the `context` and centralizes the `logging`. Everything else will follow.